File: src/kint/schemas/attendance.py
"""勤怠スキーマ。"""

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel


class AttendanceRecord(BaseModel):
    """勤怠レコード。"""

    id: str
    user_id: str
    card_idm: str | None = None
    work_date: date
    check_in: datetime | None = None
    check_out: datetime | None = None
    source: Literal["webusb_nfc", "web_user_id", "admin_manual", "self_service"]
    updated_reason: str | None = None
    is_auto_completed: bool = False
    auto_completed_at: datetime | None = None
    last_updated_at: datetime | None = None
    last_updated_by_user_id: str | None = None

    model_config = {"from_attributes": True}


class AttendanceCreateRequest(BaseModel):
    """管理者による勤怠記録追加リクエスト。"""

    user_id: str
    work_date: date
    check_in: datetime | None = None
    check_out: datetime | None = None
    reason: str


class AttendancePatchRequest(BaseModel):
    """勤怠修正リクエスト。reason は必須。"""

    check_in: datetime | None = None
    check_out: datetime | None = None
    reason: str


class AttendanceHistorySnapshot(BaseModel):
    """変更前後のスナップショット。"""

    check_in: datetime | None = None
    check_out: datetime | None = None


class AttendanceHistoryEntry(BaseModel):
    """勤怠変更履歴の1件。"""

    id: str
    attendance_id: str
    actor_user_id: str
    actor_role: Literal["admin", "employee"]
    changed_at: datetime
    before: AttendanceHistorySnapshot
    after: AttendanceHistorySnapshot
    reason: str

    model_config = {"from_attributes": True}


class AttendanceListResponse(BaseModel):
    """勤怠一覧レスポンス。"""

    items: list[AttendanceRecord]
    page: int
    page_size: int
    total: int


class AttendanceHistoryResponse(BaseModel):
    """勤怠変更履歴レスポンス。"""

    items: list[AttendanceHistoryEntry]
    total: int


class AttendanceMonthlySummary(BaseModel):
    """月次勤怠サマリー。"""

    user_id: str
    user_name: str
    full_name: str
    prescribed_days: int
    working_days: int
    total_working_hours: float
    total_overtime_hours: float
    late_count: int
    early_leave_count: int
    absence_days: int
    incomplete_days: int


class PunchPeriod(BaseModel):
    """1回分の出退勤打刻ペア。"""

    attendance_id: str | None = None
    check_in: datetime | None = None
    check_out: datetime | None = None
    calculated_check_in: datetime | None = None
    calculated_check_out: datetime | None = None
    source: str | None = None


class DailyAttendanceDetail(BaseModel):
    """日次勤怠詳細。"""

    work_date: date
    attendance_id: str | None = None
    has_shift: bool
    is_holiday: bool
    shift_start: datetime | None = None
    shift_end: datetime | None = None
    check_in: datetime | None = None
    check_out: datetime | None = None
    calculated_check_in: datetime | None = None
    calculated_check_out: datetime | None = None
    working_hours: float | None = None
    overtime_hours: float | None = None
    status: Literal[
        "normal",
        "late",
        "early_leave",
        "late_and_early",
        "absence",
        "incomplete",
        "off_duty",
    ]
    source: str | None = None
    is_auto_completed: bool = False
    punches: list[PunchPeriod] = []


class AttendanceMonthlyDetailResponse(BaseModel):
    """月間日別勤怠詳細レスポンス。"""

    user_id: str
    year_month: str
    summary: AttendanceMonthlySummary
    days: list[DailyAttendanceDetail]
    is_locked: bool = False


class AttendanceCorrectionRequestCreate(BaseModel):
    """勤怠修正申請作成リクエスト。"""

    attendance_id: str
    requested_check_in: datetime | None = None
    requested_check_out: datetime | None = None
    reason: str


class AttendanceCorrectionRequestResponse(BaseModel):
    """勤怠修正申請レスポンス。"""

    id: str
    attendance_id: str
    user_id: str
    requested_check_in: datetime | None = None
    requested_check_out: datetime | None = None
    reason: str
    status: Literal["pending", "approved", "rejected"]
    approved_by_user_id: str | None = None
    approval_comment: str | None = None
    created_at: datetime
    updated_at: datetime
    user_name: str | None = None
    user_full_name: str | None = None
    approved_by_name: str | None = None
    work_date: date | None = None
    original_check_in: datetime | None = None
    original_check_out: datetime | None = None

    model_config = {"from_attributes": True}


class AttendanceCorrectionRequestListResponse(BaseModel):
    """勤怠修正申請一覧。"""

    items: list[AttendanceCorrectionRequestResponse]
    total: int


class AttendanceCorrectionRequestApprove(BaseModel):
    """申請承認リクエスト。"""

    approval_comment: str | None = None


class AttendanceCorrectionRequestReject(BaseModel):
    """申請却下リクエスト（理由必須）。"""

    approval_comment: str


class AttendanceLockRequest(BaseModel):
    """勤怠締め処理リクエスト。"""

    year_month: str


class AttendanceLockResponse(BaseModel):
    """勤怠締め処理レスポンス。"""

    year_month: str
    locked_at: datetime
    locked_by_user_id: str

    model_config = {"from_attributes": True}

File: frontend/src/types/attendance.ts
export interface AttendanceMonthlySummary {
  user_id: string;
  user_name: string;
  full_name: string;
  email: string;
  prescribed_days: number;
  working_days: number;
  total_working_hours: number;
  total_overtime_hours: number;
  late_count: number;
  early_leave_count: number;
  absence_days: number;
  incomplete_days: number;
  yearly_working_hours: number;
}

export type DailyAttendanceStatus =
  | 'normal'
  | 'late'
  | 'early_leave'
  | 'late_and_early'
  | 'absence'
  | 'incomplete'
  | 'scheduled'
  | 'off_duty';

export interface PunchPeriod {
  attendance_id?: string | null;
  check_in: string | null;
  check_out: string | null;
  calculated_check_in?: string | null;
  calculated_check_out?: string | null;
  source?: string | null;
  overtime_reason?: string | null;
  is_manual_work_time?: boolean;
}

export interface DailyAttendanceDetail {
  work_date: string; // ISO 8601 YYYY-MM-DD
  attendance_id: string | null;
  has_shift: boolean;
  is_holiday: boolean;
  shift_start: string | null; // ISO 8601
  shift_end: string | null; // ISO 8601
  check_in: string | null; // ISO 8601
  check_out: string | null; // ISO 8601
  calculated_check_in: string | null; // ISO 8601
  calculated_check_out: string | null; // ISO 8601
  working_hours: number | null;
  overtime_hours: number | null;
  status: DailyAttendanceStatus;
  source: string | null;
  is_auto_completed?: boolean;
  is_manual_work_time?: boolean;
  punches?: PunchPeriod[];
  shifts?: { start_time: string; end_time: string }[];
}

export interface AttendanceMonthlyDetailResponse {
  user_id: string;
  year_month: string;
  summary: AttendanceMonthlySummary;
  days: DailyAttendanceDetail[];
  is_locked: boolean;
}

export interface AttendanceCorrectionRequest {
  id: string;
  attendance_id: string;
  user_id: string;
  requested_check_in: string | null;
  requested_check_out: string | null;
  reason: string;
  status: 'pending' | 'approved' | 'rejected';
  approved_by_user_id: string | null;
  approval_comment: string | null;
  created_at: string;
  updated_at: string;
  user_name: string | null;
  user_full_name: string | null;
  approved_by_name: string | null;
  work_date: string | null;
  original_check_in?: string | null;
  original_check_out?: string | null;
}

export interface AttendanceCorrectionRequestListResponse {
  items: AttendanceCorrectionRequest[];
  total: number;
}

export interface AttendanceLock {
  year_month: string;
  locked_by_user_id: string;
  locked_at: string;
}

export interface AttendanceHistorySnapshot {
  check_in: string | null;
  check_out: string | null;
  work_start: string | null;
  work_end: string | null;
}

export interface AttendanceHistoryEntry {
  id: string;
  attendance_id: string;
  actor_user_id: string;
  actor_role: 'admin' | 'employee';
  changed_at: string;
  before: AttendanceHistorySnapshot;
  after: AttendanceHistorySnapshot;
  reason: string;
}

export interface AttendanceHistoryResponse {
  items: AttendanceHistoryEntry[];
  total: number;
}

export interface AttendanceRecord {
  id: string;
  user_id: string;
  card_idm: string | null;
  work_date: string;
  check_in: string | null;
  check_out: string | null;
  work_start: string | null;
  work_end: string | null;
  is_manual_work_time: boolean;
  source: 'webusb_nfc' | 'web_user_id' | 'admin_manual' | 'self_service';
  updated_reason: string | null;
  is_auto_completed: boolean;
  auto_completed_at: string | null;
  last_updated_at: string | null;
  last_updated_by_user_id: string | null;
  overtime_reason?: string | null;
}

